### PR TITLE
Disable validation of field lengths when parsing event

### PR DESCRIPTION
**Pull request by istreeter**

### CODE REVIEW

File: .github/workflows/test_and_deploy.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ jobs:
           java-version: 11
 
       - name: Run tests
-        run: sbt coverage +test
+        run: sbt 'set coverageEnabled := true' +test
```
Comment (istreeter) [inline .github/workflows/test_and_deploy.yml:18]: I don't know why this change was needed, but the github action failed for scala3 without this change.

File: src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala [hunk 2]
```diff
@@ -242,30 +243,56 @@ case class Event(
...
   /**
    * Derived TSV parser for the Event class
    */
-  private val parser: Parser[Event] = Parser.deriveFor[Event].get
+  private lazy val parser: Parser[Event] = Parser.deriveFor[Event].get(Map.empty)
+
+  /**
+   * Derived TSV parser for the Event class
+   */
+  private lazy val parserWithValidation: Parser[Event] = Parser.deriveFor[Event].get(FIELD_SIZES)
```
Comment (benjben) [inline src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala:266]: Good idea to use `lazy` :+1:

**Review by benjben — APPROVED** [review]
Looks great !

File: src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/decode/ValueDecoder.scala [hunk 3]
```diff
@@ -40,42 +40,49 @@ import com.snowplowanalytics.snowplow.analytics.scalasdk.ParsingError.RowDecodin
...
     fromFunc[String] {
-      case (key, value) =>
-        if (value.length > FIELD_SIZES.getOrElse(key.name, Int.MaxValue))
-          InvalidValue(key,
-                       value,
-                       s"Field ${key.name} longer than maximum allowed size ${FIELD_SIZES.getOrElse(key.name, Int.MaxValue)}"
-          ).asLeft
-        else if (value.isEmpty) InvalidValue(key, value, s"Field ${key.name} cannot be empty").asLeft
-        else value.asRight
+      case (key, value, Some(maxLength)) if value.length > maxLength =>
```
[the remark this answers is not in the document]
Comment (istreeter) [reply src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/decode/ValueDecoder.scala:64]: Version 3.0.0 used it [on this line](https://github.com/snowplow/snowplow-scala-analytics-sdk/blob/3.0.1/src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala#L254).  So it's a bit like my other comment -- we need to keep it around so that we maintain binary compatibility with previous 3.x versions.

**Comment by istreeter** [comment]
Thanks @benjben.  I pushed an extra commit to address some of your ideas, particularly in the tests.

File: src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala [hunk 4]
```diff
@@ -242,30 +243,56 @@ case class Event(
 
 object Event {
 
+  @deprecated("Event.unsafe functionality is merged into Event.parse method", "3.1.0")
   object unsafe {
-    implicit def unsafeEventDecoder: Decoder[Event] = deriveDecoder[Event]
+    implicit def unsafeEventDecoder: Decoder[Event] = Event.eventDecoder
   }
```
[the remark this answers is not in the document]
Comment (benjben) [reply src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala:249]: I see, thanks for the explanation.

> An alternative option would be to release this as version 4.0.0 and then break binary compatibility. But I figured Snowplow users would probably prefer to minimize this kind of disruption.

Indeed. Although it will be weird to release 4.0.0 with this legacy code in it when the time comes (unless we think of removing it, but then why not releasing 4.0.0 now).

**Review by benjben — APPROVED** [review]
:+1:

**Review by benjben — APPROVED** [review]
